docs/developer-docs/security/security-best-practices/inter-canister-calls.mdx

```diff
@@ -0,0 +1,330 @@
+    // and locking would not be effective
+    let _guard = CallerGuard::new(caller)?;
+    // do anything, call other canisters
+    Ok(())
+} // here the guard goes out of scope and is dropped
+
+mod test {
+    use super::*;
+
+    #[test]
+    fn should_obtain_guard_for_different_principals() {
+        let principal_1 = Principal::anonymous();
+        let principal_2 = Principal::management_canister();
+        let caller_guard = CallerGuard::new(principal_1);
+        assert!(caller_guard.is_ok());
+        assert!(CallerGuard::new(principal_2).is_ok());
+    }
+
+    #[test]
+    fn should_not_obtain_guard_twice_for_same_principal() {
+        let principal = Principal::anonymous();
+        let caller_guard = CallerGuard::new(principal);
+        assert!(caller_guard.is_ok());
+        assert!(CallerGuard::new(principal).is_err());
+    }
+
+    #[test]
+    fn should_release_guard_on_drop() {
+        let principal = Principal::anonymous();
+        {
+            let caller_guard = CallerGuard::new(principal);
+            assert!(caller_guard.is_ok());
+        } // drop caller_guard as it goes out of scope here
+        // it is possible to get a guard again:
+        assert!(CallerGuard::new(principal).is_ok());
+    }
+}
+```
+
+This pattern can be extended e.g. to work for the following use cases:
+
+- A global lock that does not only lock per caller. For this, set a boolean flag in the canister state instead of using a `BTreeSet<Principal>`.
+- A guard that makes sure that only a limited number of principals are allowed to execute a method at the same time. For this, one can return an error in `CallerGuard::new()` in case `pending_requests.len() >= MAX_NUM_CONCURRENT_REQUESTS`.
+- A guard that limits the number of times a method can be called in parallel. For this, use a counter in the canister state that is checked and increased in `CallerGuard::new()` and decreased in `Drop`.
+- A guard that makes sure that every task from a set of tasks can only be processed once, independently of the caller who triggered the processing. [View example project](https://github.com/dfinity/examples/tree/master/rust/guards).
+- A lock that uses a different type than `Principal` to grant access to the resource. [View an implementation using generic types](./_attachments/lock.rs).
+
+Finally, note that the same guard can be used in several methods to restrict parallel execution of them.
+
+## Handle rejected inter-canister calls correctly
+
+### Security concern
+
+As stated by the Property 6 above, inter-canister calls can fail in which case they result in a **reject**. See [reject codes](/docs/current/references/ic-interface-spec#reject-codes) for more detail. The caller must correctly deal with the reject cases, as they can happen in normal operation, e.g. because of insufficient cycles on the sender or receiver side, or because some data structures (like message queues) are full.
```

Comment:
```suggestion
As stated by the Property 6 above, inter-canister calls can fail in which case they result in a **reject**. See [reject codes](/docs/current/references/ic-interface-spec#reject-codes) for more detail. The caller must correctly deal with the reject cases, as they can happen in normal operation, because of insufficient cycles on the sender or receiver side, or because some data structures like message queues are full.
```

Reply:
Done